internal/controller/grafanaorganization_controller.go

```diff
@@ -149,173 +148,34 @@ func (r GrafanaOrganizationReconciler) reconcileCreate(ctx context.Context, graf
-	}
-
-	return nil
-}
-
-func (r GrafanaOrganizationReconciler) configureDatasources(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, grafanaOrganization *v1alpha1.GrafanaOrganization) error {
-	logger := log.FromContext(ctx)
-
-	logger.Info("configuring data sources")
+	service := grafana.NewService(grafanaAPI)
 
-	// Create or update organization in Grafana
-	var organization = newOrganization(grafanaOrganization)
-	datasources, err := grafana.ConfigureDefaultDatasources(ctx, grafanaAPI, organization)
+	err = service.SetupOrganization(ctx, grafanaOrganization)
 	if err != nil {
-		logger.Error(err, "failed to configure the grafanaOrganization with default datasources")
-		return errors.WithStack(err)
-	}
-
-	var configuredDatasources = make([]v1alpha1.DataSource, len(datasources))
-	for i, datasource := range datasources {
-		configuredDatasources[i] = v1alpha1.DataSource{
-			ID:   datasource.ID,
-			Name: datasource.Name,
-		}
-	}
-
-	logger.Info("updating datasources in the grafanaOrganization status")
-	grafanaOrganization.Status.DataSources = configuredDatasources
-	if err := r.Status().Update(ctx, grafanaOrganization); err != nil {
-		logger.Error(err, "failed to update the the grafanaOrganization status with datasources information")
-		return errors.WithStack(err)
+		return ctrl.Result{}, errors.WithStack(err)
 	}
-	logger.Info("updated datasources in the grafanaOrganization status")
-	logger.Info("configured data sources")
 
-	return nil
+	return ctrl.Result{}, nil
 }
 
 // reconcileDelete deletes the grafana organization.
 func (r GrafanaOrganizationReconciler) reconcileDelete(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, grafanaOrganization *v1alpha1.GrafanaOrganization) error {
-	logger := log.FromContext(ctx)
 
 	// We do not need to delete anything if there is no finalizer on the grafana organization
 	if !controllerutil.ContainsFinalizer(grafanaOrganization, v1alpha1.GrafanaOrganizationFinalizer) {
 		return nil
 	}
 
-	// Delete organization in Grafana if it exists
-	var organization = newOrganization(grafanaOrganization)
-	if grafanaOrganization.Status.OrgID > 0 {
-		err := grafana.DeleteOrganization(ctx, grafanaAPI, organization)
-		if err != nil {
-			return errors.WithStack(err)
-		}
+	service := grafana.NewService(grafanaAPI)
 
-		grafanaOrganization.Status.OrgID = 0
-		if err = r.Status().Update(ctx, grafanaOrganization); err != nil {
-			logger.Error(err, "failed to update grafanaOrganization status")
-			return errors.WithStack(err)
-		}
-	}
-
-	var errs []error
-
-	// Configure Grafana RBAC
-	if err := r.configureGrafanaSSO(ctx, grafanaAPI); err != nil {
-		errs = append(errs, err)
-	}
-
-	if len(errs) > 0 {
-		return errors.WithStack(stderrors.Join(errs...))
-	}
-
-	// Finalizer handling needs to come last.
-	err := r.finalizerHelper.EnsureRemoved(ctx, grafanaOrganization)
+	err := service.UnsetupOrganization(ctx, grafanaOrganization)
```

Comment:
You're forgetting about the status updates here right? Those need to stay in the controller

Reply:
The grafana service has the client to do the necessary status update